lib/schedule/swing.ex

```diff
@@ -0,0 +1,88 @@
+defmodule Schedule.Swing do
+  alias Schedule.Hastus
+  alias Schedule.Gtfs.Service
+  alias Schedule.Route
+  alias Schedule.Trip
+  alias Schedule.Minischedule
+
+  @type t :: %__MODULE__{
+          from_route_id: Route.id(),
+          from_run_id: Hastus.Run.id(),
+          from_trip_id: Trip.id(),
+          to_route_id: Route.id(),
+          to_run_id: Hastus.Run.id(),
+          to_trip_id: Trip.id(),
+          time: Util.Time.time_of_day()
+        }
+
+  @derive Jason.Encoder
+
+  defstruct [
+    :from_route_id,
+    :from_run_id,
+    :from_trip_id,
+    :to_route_id,
+    :to_run_id,
+    :to_trip_id,
+    :time
+  ]
+
+  @type by_schedule_id_and_route_id :: %{{Service.id(), Route.id()} => [t()]}
+
+  @spec from_minischedule_blocks(Minischedule.Block.by_id(), Trip.by_id()) ::
+          by_schedule_id_and_route_id()
+  def from_minischedule_blocks(minischedule_blocks, trips_by_id) do
```

Comment:
this function is a bit difficult to follow due to its bulk, I suggest extracting some helper functions here